_source/_posts/2021-12-06-spring-data-elasticsearch.md

```diff
@@ -0,0 +1,349 @@
+
+```json
+{
+  "mappings": {
+    "_doc": {
+      "properties": {
+        "_class": {
+          "type": "keyword",
+          "index": false,
+          "doc_values": false
+        },
+        "id": {
+          "type": "text",
+          "fields": {
+            "keyword": {
+              "type": "keyword",
+              "ignore_above": 256
+            }
+          }
+        },
+        "name": {
+          "type": "text",
+          "fields": {
+            "keyword": {
+              "type": "keyword",
+              "ignore_above": 256
+            }
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+Now the `id` and `name` properties have been _dynamically mapped_ with _multi-field mapping_. For example, the `name` field is defined with type `text`, for full-text searches, and it defines a `name.keyword` sub-field for aggregations and sorting:
+
+```json
+"name": {
+  "type": "text",
+  "fields": {
+    "keyword": {
+      "type": "keyword",
+      "ignore_above": 256
+    }
+  }
+}
+```
+
+It is often useful to index the same field in different ways for different purposes. Text fields are searchable by default, but by default are not available for aggregations, sorting, unless multi-field mapping is in place. There is another option for enabling sorting for text fields, which is the [`fielddata` mapping parameter](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#fielddata-mapping-param), which will not be covered in this tutorial, as it is discouraged in the Elasticsearch documentation, because of high memory consumption.
+
+
+Elasticsearch exposes several APIs, and the Index APIs allow managing individual indices, index settings, aliases, mappings, and index templates. Although Kibana is a user-friendly front-end, you can also query the index mappings through the Index APIs, with an HTTPie command like the following:
+
+```shell
+http localhost:9200/tag/_mapping
+```
+
+The output should look the same as the mapping visualized in Kibana.
+
+## Key components in reactive Spring Data Elasticsearch
+
+As mentioned before, the Blog application is reactive, and that was defined in the JDL file with `reactive true` in the `gateway` and `blog` application configuration. That also signals JHipster generators to include the reactive Elasticsearch dependencies in the `blog` microservice. JHipster generates the web layer with the entity resource class, and it optionally generates the service class for abstracting from the data layer. Two of the key components that enable the Elasticsearch reactive integration are the `ReactiveElasticsearchClient` and the `ReactiveElasticsearchOperations`. A client is required to connect to an Elasticsearch cluster or node, and operations is the interface for the execution of CRUD and query commands. The third component is the entity `ReactiveElasticsearchRepository`, which can provide domain-specific search methods. Let's briefly describe each of these components.
```

Comment:
```suggestion
As mentioned before, the `blog` application is reactive, and that's because the JDL file had `reactive true` in its `gateway` and `blog` application configurations. 

This config also tels JHipster to include the Elasticsearch dependencies in the `blog` microservice. JHipster generates the web layer with the entity resource class, and it optionally generates the service class too. Two of the key components that enable the Elasticsearch reactive integration are the `ReactiveElasticsearchClient` and the `ReactiveElasticsearchOperations`. A client is required to connect to an Elasticsearch cluster or node, and `*Operations` is the interface for the execution of CRUD and query commands. The third component is the entity `ReactiveElasticsearchRepository`, which can provide domain-specific search methods. Let's briefly describe each of these components.
```